bundle/jsky.app.ot/src/main/scala/jsky/app/ot/ags/BagsManager.scala

```diff
@@ -0,0 +1,321 @@
+        taskMap.get(obsKey).fold(true){t =>
+          val done = t.isDone
+          if (done) taskMap -= obsKey
+          done
+        }
+      }
+      if (!doneWithTask)
+        taskQueue.synchronized {
+          if (!taskQueue.contains(obsKey))
+            taskQueue.add(obsKey)
+        }
+    }
+    else {
+      Thread.sleep(RequeueDelay)
+
+      // We only need to do something if the task is not already in the queue.
+      taskQueue.synchronized {
+        if (!taskQueue.contains(obsKey)) {
+          val newTask = task match {
+            case BagsTask(obs, Some(ctx), None) => BagsTask(obs, None, Some(ctx))
+            case t => t
+          }
+          taskMap.synchronized(taskMap += ((obsKey, newTask)))
+          taskQueue.add(obsKey)
+        }
+      }
+    }
+  }
+
+  // Given an observation, either insert it in all data structures or update the existing entry for it.
+  private def updateObservation(obs: ISPObservation) =
+    Option(obs).foreach { ispObs =>
+      val obsKey = ispObs.getNodeKey
+      val ctxOpt = ObsContext.create(ispObs).asScalaOpt
+
+      // Update the task map and the task queue.
+      taskMap.synchronized {
+        val task = taskMap.get(obsKey) match {
+          case Some(t) => t.newContext(ctxOpt)
+          case None    => BagsTask(ispObs, None, ctxOpt)
+        }
+        taskMap += ((obsKey, task))
+      }
+      taskQueue.synchronized {
+        if (!taskQueue.contains(obsKey))
+          taskQueue.add(obsKey)
+      }
+    }
+
+  // Remove a given observation from all data structures.
+  private def removeObservation(obs: ISPObservation) = {
+    Option(obs).foreach { ispObs =>
+      val obsKey = ispObs.getNodeKey
+      taskQueue.synchronized(taskQueue.remove(obsKey))
+      taskMap.synchronized(taskMap -= obsKey)
+    }
+  }
+
+  // Check two target environments to see if the BAGS targets match exactly between them.
+  def bagsTargetsMatch(oldEnv: TargetEnvironment, newEnv: TargetEnvironment): Boolean =
+    oldEnv.getGuideEnvironment.getPrimaryReferencedGuiders.equals(newEnv.getGuideEnvironment.getPrimaryReferencedGuiders) &&
+      oldEnv.getGuideEnvironment.getPrimaryReferencedGuiders.asScala.forall { gp =>
+        oldEnv.getPrimaryGuideProbeTargets(gp).asScalaOpt.forall { oldGpt =>
+          val oldBags = oldGpt.getBagsTarget.asScalaOpt
+          val newBags = newEnv.getPrimaryGuideProbeTargets(gp).asScalaOpt.flatMap(_.getBagsTarget.asScalaOpt)
+          (oldBags.isEmpty && newBags.isEmpty) || oldBags.exists(oldTarget => newBags.exists(_.getTarget.equals(oldTarget.getTarget)))
+        }
+      }
+
+  private val propertyChangeListener = new PropertyChangeListener {
+    override def propertyChange(evt: PropertyChangeEvent): Unit = {
+      evt.getSource match {
+        case obsComp: ISPObsComponent =>
+          // Check to see if anything has changed.
+          val sameEnv = evt.getOldValue.isInstanceOf[TargetObsComp] && evt.getNewValue.isInstanceOf[TargetObsComp] && {
```

Comment:
I'm not sure how to do this or if it's possible? Since these are of type `AnyRef`, I think there may be complaints from the compiler about type patterns being unchecked / eliminated by erasure?

Or maybe I have no idea what I am talking about :D.